### PR TITLE
Simulator integration with FCI Cellnet

### PR DESCRIPTION
**Pull request by yhwen**
### Description

Simulator integration using the FCI cellnet communication.
 

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/private/fed/simulator/simulator_app_runner.py [hunk 1]
```diff
@@ -24,8 +24,13 @@ def create_job_processing_context_properties(self, workspace, job_id):
 
 
 class SimulatorClientAppRunner(ClientAppRunner):
+    # def start_command_agent(self, args, client_runner, federated_client, fl_ctx):
+    #     super().start_command_agent(args, client_runner, federated_client, fl_ctx)
+
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/simulator/simulator_app_runner.py:29]: we should remove commented out codes.

File: nvflare/private/fed/simulator/simulator_server.py [hunk 2]
```diff
@@ -134,3 +134,15 @@ def create_run_manager(self, workspace, job_id):
             client_manager=self.client_manager,
             handlers=self.runner_config.handlers,
         )
+
+    def wait_engine_run_complete(self, name, shutdown_grace_time=2.0, cleanup_grace_time=3.0):
+        while self.engine.engine_info.status != MachineStatus.STOPPED:
+            if self.engine.asked_to_stop:
+                self.engine.engine_info.status = MachineStatus.STOPPED
+
+            time.sleep(3)
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/simulator/simulator_server.py:143]: this should use the argument `cleanup_grace_time` ?

File: nvflare/private/fed/app/simulator/simulator_runner.py [hunk 3]
```diff
@@ -130,7 +133,8 @@ def setup(self):
         os.chdir(self.args.workspace)
         fobs_initialize()
         AuthorizationService.initialize(EmptyAuthorizer())
-        AuditService.initialize(audit_file_name=WorkspaceConstants.AUDIT_LOG)
+        # AuditService.initialize(audit_file_name=WorkspaceConstants.AUDIT_LOG)
+        AuditService.the_auditor = SimulatorAuditor()
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/app/simulator/simulator_runner.py:137]: Why do we need this `SimulatorAuditor` for?
Are we just mocking all its functionality?

File: nvflare/private/fed/simulator/simulator_audit.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvflare.fuel.sec.audit import Auditor
+
+
+class SimulatorAuditor(Auditor):
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/simulator/simulator_audit.py:18]: Do we need this class?

File: nvflare/private/fed/app/simulator/simulator_runner.py [hunk 5]
```diff
@@ -467,11 +530,14 @@ def do_one_task(self, client, num_of_threads, gpu, lock):
...
     def _create_connection(self, open_port):
         conn = None
+        start = time.time()
         while not conn:
             try:
                 address = ("localhost", open_port)
                 conn = Client(address, authkey=CommunicationMetaData.CHILD_PASSWORD.encode())
             except BaseException:
+                if time.time() - start > 60.0:
+                    raise RuntimeError(f"Failed to create connection to the child process in {self.__class__.__name__}")
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/app/simulator/simulator_runner.py:540]: We might want to make this 60 seconds configurable.

And we should include this 60 seconds timeout in the `RuntimeError`?

`f"Failed to create connection to the child process in {self.__class__.__name__} in {timeout} seconds"`

File: nvflare/private/fed/simulator/simulator_client_engine.py [hunk 6]
```diff
@@ -12,26 +12,29 @@
...
-        super().__init__(client, client_name, sender, args, rank, workers)
-
-    def send_aux_command(self, shareable: Shareable, job_id):
-        run_manager = self.client.run_manager
-        if run_manager:
-            with run_manager.new_context() as fl_ctx:
-                topic = shareable.get_header(ReservedHeaderKey.TOPIC)
-                return run_manager.dispatch(topic=topic, request=shareable, fl_ctx=fl_ctx)
+    def __init__(self, client, client_name, args, rank, workers=5):
+        super().__init__(client, client_name, args, rank, workers)
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/simulator/simulator_client_engine.py:23]: If this is the same as `ClientEngine`, let's get rid of this class `SimulatorClientEngine`

File: nvflare/private/fed/app/simulator/simulator.py [hunk 7]
```diff
@@ -51,8 +52,18 @@ def run_simulator(simulator_args):
     create the SimulatorRunner object, do a setup(), then calls the run().
     """
 
-    if sys.version_info < (3, 7):
-        raise RuntimeError("Please use Python 3.7 or above.")
+    # For MacOS, it needs to use 'spawn' for creating multi-process.
+    if platform == "linux" or platform == "linux2":
```
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/private/fed/app/simulator/simulator.py:56]: Removed the Linux check.

File: nvflare/private/fed/app/simulator/simulator_runner.py [hunk 8]
```diff
@@ -130,7 +133,8 @@ def setup(self):
         os.chdir(self.args.workspace)
         fobs_initialize()
         AuthorizationService.initialize(EmptyAuthorizer())
-        AuditService.initialize(audit_file_name=WorkspaceConstants.AUDIT_LOG)
+        # AuditService.initialize(audit_file_name=WorkspaceConstants.AUDIT_LOG)
```
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/private/fed/app/simulator/simulator_runner.py:136]: Removed.

File: nvflare/private/fed/app/simulator/simulator_runner.py [hunk 9]
```diff
@@ -302,17 +308,35 @@ def run(self):
...
             self.logger.info("KeyboardInterrupt, terminate all the child processes.")
             kill_child_processes(os.getpid())
             return -9
 
     def run_processs(self, return_dict):
+        # run_status = self.simulator_run_main()
+        run_status = mpm.run(main_func=self.simulator_run_main, shutdown_grace_time=3, cleanup_grace_time=6)
+
+        return_dict["run_status"] = run_status
+        # os._exit(0)
```
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/private/fed/app/simulator/simulator_runner.py:326]: Removed.

File: nvflare/private/fed/app/simulator/simulator_runner.py [hunk 10]
```diff
@@ -302,17 +308,35 @@ def run(self):
...
+
+    def simulator_run_main(self):
         if self.setup():
             try:
                 self.create_clients()
+                # self.services.engine.run_processes[SimulatorConstants.JOB_NAME][RunProcessKey.PARTICIPANTS] =
```
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/private/fed/app/simulator/simulator_runner.py:332]: Removed.

File: nvflare/private/fed/simulator/simulator_app_runner.py (hunk 1 above)
Comment (yhwen) [reply nvflare/private/fed/simulator/simulator_app_runner.py:29]: Removed.

File: nvflare/private/fed/simulator/simulator_server.py (hunk 2 above)
Comment (yhwen) [reply nvflare/private/fed/simulator/simulator_server.py:143]: changed.

File: nvflare/private/fed/app/simulator/simulator_runner.py (hunk 3 above)
Comment (yhwen) [reply nvflare/private/fed/app/simulator/simulator_runner.py:137]: Simulator run server and clients in the same process. Each client / server share the same AuditService, which will write the audit log in a file. large number of concurrent clients will cause a file write error. Simulator does not need the audit log. SimulatorAuditor mock the function.

File: nvflare/private/fed/simulator/simulator_audit.py (hunk 4 above)
Comment (yhwen) [reply nvflare/private/fed/simulator/simulator_audit.py:18]: Mock the Auditor functions.

File: nvflare/private/fed/app/simulator/simulator_runner.py (hunk 5 above)
Comment (yhwen) [reply nvflare/private/fed/app/simulator/simulator_runner.py:540]: Done.

File: nvflare/private/fed/simulator/simulator_client_engine.py (hunk 6 above)
Comment (yhwen) [reply nvflare/private/fed/simulator/simulator_client_engine.py:23]: Done.

File: nvflare/private/fed/simulator/simulator_app_runner.py [hunk 11]
```diff
@@ -25,7 +25,9 @@ def create_job_processing_context_properties(self, workspace, job_id):
 
 class SimulatorClientAppRunner(ClientAppRunner):
     def start_command_agent(self, args, client_runner, federated_client, fl_ctx):
-        pass
+        # pass
+        super().start_command_agent(args, client_runner, federated_client, fl_ctx)
+        self.command_agent.register_cell_cb()
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/simulator/simulator_app_runner.py:30]: In the `ClientAppRunner`'s `start_command_agent` we call `self.command_agent.start(fl_ctx)`.

And in that `start` function we already have `self.register_cell_cb()`

So we don't need this method anymore?

File: nvflare/private/fed/simulator/simulator_server.py [hunk 12]
```diff
@@ -134,3 +126,15 @@ def create_run_manager(self, workspace, job_id):
             client_manager=self.client_manager,
             handlers=self.runner_config.handlers,
         )
+
+    def wait_engine_run_complete(self, name, shutdown_grace_time=2.0, cleanup_grace_time=3.0):
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/simulator/simulator_server.py:130]: arg `shutdown_grace_time` is not used, we can remove it

Comment (yhwen) [reply nvflare/private/fed/simulator/simulator_server.py:130]: Don't need this wait_engine_run_complete() any more. Deleted.

File: nvflare/private/fed/app/simulator/simulator_runner.py [hunk 13]
```diff
@@ -417,35 +470,45 @@ def run_client_thread(self, num_of_threads, gpu, lock):
...
             + " --port "
             + str(open_port)
             + " --parent_pid "
             + str(os.getpid())
+            + " --simulator_root "
+            + self.simulator_root
+            + " --root_url "
+            + str(client.cell.get_root_url_for_child())
+            + " --parent_url "
+            + str(client.cell.get_internal_listener_url())
```
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/private/fed/app/simulator/simulator_runner.py:498]: this is subjective. This is easier for human readability.

**Comment by yhwen** [comment]
> can you clean up the comment out code ?

Done

**Review by YuanTingHsieh — APPROVED** [review]
Thanks, LGTM.

**Review by nvidianz — COMMENTED** [review]
LGTM

**Pull request merged**